Sign in via OAuth for new REST APIs

Envek opened the pull request:
https://developer.ebay.com/api-docs/static/oauth-details.html

What is done:

 1. Configuration (except scopes)
 2. Request phase: URL
 3. Callback phase: getting access token
 4. Callback phase: getting user info from old APIs to populate structures required by OmniAuth (and applications): see https://forums.developer.ebay.com/questions/18476/i-want-to-get-user-info-with-new-rest-api-like-get.html for details (no new APIs for that yet)
 5. Access token refreshing
 6. Partial tests

Envek committed:
Extract default Auth'n'auth method to adapter

To allow implementation of another methods

Envek committed:
Implement OAuth authentication method

Due to the fact, that there is no REST API for getting logged in user's
information, so old Trading API GetUser call is used instead.
See: https://forums.developer.ebay.com/questions/18476/i-want-to-get-user-info-with-new-rest-api-like-get.html

Envek committed:
Allow to use new OAuth access tokens to access old XML APIs

Envek committed:
Specs and documentation for TokenSet class

Envek committed:
Allow to use old and new access tokens simultaneously

Prefer new access tokens if available

Envek committed:
Make RuboCop a bit happier

nepalez commented on lib/ebay_request/auth.rb:
```diff
@@ -1,35 +1,24 @@
-    ruparams = CGI.escape(ruparams.map { |k, v| "#{k}=#{v}" }.join("&"))
-    params << "ruparams=#{CGI.escape(ruparams)}"
+class EbayRequest::Auth
+  extend Dry::Initializer::Mixin
```
Dry::Initializer::Mixin deprecated
Лучше просто Dry::Initializer

nepalez replied:
И еще вопрос, а этот функционал точно нужен лишь в Trading? Будет ли он работать для других частей апи?

Envek replied:
Похоже, что из старых апишек доступ по токену требуется только в Trading API, остальные — публичные.

nepalez commented on lib/ebay_request/auth/auth_n_auth.rb:
```diff
@@ -0,0 +1,79 @@
+  end
+
+  def full_name
+    @full_name ||= raw_info["RegistrationAddress"].try(:[], "Name")
```
try ? Рельсовый? ;)

nepalez replied:
raw_info.dig("RegistrationAddress", "Name").to_s

Envek replied:
Ох, стыд >_<
Исправил

nepalez commented on lib/ebay_request/auth/auth_n_auth.rb:
```diff
@@ -0,0 +1,79 @@
+  private
+
+  def parsed_name
+    @parsed_name ||= (full_name || "").split(" ", 2)
```
full_name.split ...

nepalez commented on lib/ebay_request/auth/oauth.rb:
```diff
@@ -0,0 +1,123 @@
+      first_name: parsed_name[0],
+      last_name:  parsed_name[1],
+      eias_token: raw_info["EIASToken"],
+      country:    raw_info["RegistrationAddress"].try(:[], "Country"),
```
Мне таки кажется, что о try нужно забыть даже в Rails
&.
В хэше -- dig

nepalez commented on lib/ebay_request/auth/oauth.rb:
```diff
@@ -0,0 +1,123 @@
+  end
+
+  def parsed_name
+    @parsed_name ||= (full_name || "").split(" ", 2)
```
full_name лучше сделать строкой всегда -- см выше

nepalez commented on lib/ebay_request/token_set.rb:
```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+require "httparty"
```
Вот этот кусок мне кажется неправильным
Зачем добавлять лишнюю зависимость?

Envek replied:
Убрал, но Net::HTTP — это боль.

Envek committed:
Remove rails-specific code from authentication adapters

Envek committed:
Replace Dry::Initializer::Mixin with Dry::Initializer everywhere

Envek committed:
Get rid of httparty and use Net::HTTP to get tokens

Envek committed:
Allow to specify required OAuth scopes (use all by default)

Envek committed:
Rename TokenSet to IAFTokenManager

To clearly distinguish new tokens got from OAuth
from old tokens got from Auth'n'auth.

Envek committed:
Refresh IAF token automatically if expired error is returned from API

nepalez committed:
Fix error to be risen by `EbayRequest::Trading#request`

Unlike `EbayRequest::Response#error_class`, the `EbayRequest::Response#error` is an istance of error_class, that contains a hash of error messages assotiated to their codes.

Knowledge of those codes is essential for processing errors correctly (by the app that can not be called)

nepalez committed:
Fix cases when token expiration times are empty (#2)

Envek commented:
OmniAuth strategy for new APIs (OAuth) is now at [omniauth-ebay-oauth](https://github.com/evilmartians/omniauth-ebay-oauth) gem
IAF token manager class is at [ebay_api](https://github.com/nepalez/ebay_api/) client gem for eBay REST APIs
The only thing left for ebay_request is support for calling this token manager, so I will extract it to separate pull request.
